streaming/base/partitioning.py

```diff
@@ -81,19 +80,94 @@ def get_partitions(num_samples: int,
         last = np.full((num_physical_nodes, too_many, ranks_per_node), -1, np.int64)
         x = np.concatenate([x, last], 1)
 
     # -> Shape: (physical nodes x ranks per node x samples)
     x = x.transpose(0, 2, 1)
-    x = x.reshape(num_physical_nodes, ranks_per_node, -1, batch_size)
+    x = x.reshape(num_physical_nodes, ranks_per_node, -1, device_batch_size)
     x = np.concatenate(
         [x[:, :, np.arange(i, x.shape[2], workers_per_rank), :] for i in range(workers_per_rank)],
         axis=2)
 
     # -> Shape: (physical nodes x ranks per node x workers per rank x samples)
     return x.reshape(num_physical_nodes, ranks_per_node, workers_per_rank, -1)
+
+
+def get_partitions_fast(dataset_size: int,
+                        num_canonical_nodes: int,
+                        num_physical_nodes: int,
+                        ranks_per_node: int,
+                        workers_per_rank: int,
+                        device_batch_size: int = 1,
+                        drop_first: int = 0):
+    """Partition the given number of samples to nodes, devices, and workers.
+
+    Args:
+        dataset_size (int): Dataset size.
+        num_canonical_nodes (int): Number of canonical nodes.
+        num_physical_nodes (int): Number of physical nodes.
+        ranks_per_node (int): Number of ranks per node.
+        workers_per_rank (int): Number of worker partitions per rank.
+        device_batch_size (int): Batch size of its DataLoader, which affects how the dataset is
+            partitioned over the workers. Defaults to ``1``.
+        drop_first (int): Number of samples seen already, which are dropped. Defaults to ``0``.
+
+    Returns:
+        NDArray[np.int64]: Partitions of shape (physical nodes x ranks x workers x samples).
+    """
+    assert not num_canonical_nodes % num_physical_nodes or \
+        not num_physical_nodes % num_canonical_nodes
+    device_batch_size = device_batch_size or 1
+
+    devices = num_physical_nodes * ranks_per_node
+    device_samples = math.ceil(dataset_size / devices)
+    worker_batches = math.ceil(device_samples / (workers_per_rank * device_batch_size))
+    padded_device_samples = workers_per_rank * worker_batches * device_batch_size
+
+    node_starts = dataset_size * np.arange(num_canonical_nodes) // num_canonical_nodes
+    device_starts = np.arange(ranks_per_node)
+    step = ranks_per_node
+
+    if num_canonical_nodes < num_physical_nodes:
+        node_ratio = num_physical_nodes // num_canonical_nodes
+        node_starts = node_starts.repeat(node_ratio)
+        node_starts += np.arange(node_ratio).repeat(num_canonical_nodes)
+        device_starts *= node_ratio
+        step *= node_ratio
+
+    starts = node_starts.reshape(-1, 1, 1) + device_starts.reshape(1, -1, 1)
+    indices = np.arange(padded_device_samples).reshape(1, 1, -1)
+    ids = starts + indices * step
+
+    if num_physical_nodes < num_canonical_nodes:
+        node_ratio = num_canonical_nodes // num_physical_nodes
```

Comment:
```suggestion
        # reshape the ids so that the order of the samples is preserved based on the provided number
        # of canonical nodes
        node_ratio = num_canonical_nodes // num_physical_nodes
```